### PR TITLE
GH-2029: Two bugs caused by recursive structural types

### PR DESCRIPTION
**Pull request by mor-n4**
See #2029.

### CODE REVIEW

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/utils/StructuralTypingComputer.xtend [hunk 1]
```diff
@@ -103,9 +103,10 @@ class StructuralTypingComputer extends TypeSystemHelperStrategy {
 		if (G.isStructuralSubtypingInProgressFor(left, right)) {
 			return result(left, right, emptyList, emptyList);
 		}
-		val G2 = G.wrap; // don't pollute incoming rule environment with our recursion guard
+		val G2 = G.wrapIfInitialInvocation; // don't pollute incoming rule environment with our recursion guard
 		G2.rememberStructuralSubtypingInProgressFor(left, right);
 
+//println("!!!");
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/utils/StructuralTypingComputer.xtend:109]: left over from debugging?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java [hunk 2]
```diff
@@ -731,6 +732,23 @@ private boolean reduceStructuralTypeRef(TypeRef left, TypeRef right, Variance va
 		}
 		// now, variance is either CO or INV
 
+		// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java:735]: How about introducing a method at this location?

**Review by mmews-n4 — APPROVED** [review]
👍

**Comment by mor-n4** [comment]
Thanks!

Builds green, merging ...

**Pull request merged**